### PR TITLE
[LWG 3] P2407R5 Freestanding Library: Partial Classes

### LINKED ISSUE #6661: [2023-11 LWG Motion 3] Freestanding Library: Partial Classes
https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2023/p2407r5.html

### PR DESCRIPTION
**Pull request by jensmaurer**
Fixes #6661 
Fixes cplusplus/papers#1077

### CODE REVIEW

File: source/strings.tex [hunk 1]
```diff
@@ -1520,7 +1532,7 @@
 \item
 \tcode{xpos < size()}
 \item
-\tcode{traits::eq(at(xpos), str.at(I))} for some element \tcode{I} of the string referenced by \tcode{str}.
+\tcode{traits::eq(data_[xpos], str.[I])} for some element \tcode{I} of the string referenced by \tcode{str}.
```
Comment (JohelEGP) [inline source/strings.tex:1535]:
```suggestion
\tcode{traits::eq(data_[xpos], str[I])} for some element \tcode{I} of the string referenced by \tcode{str}.
```

**Review by JohelEGP — CHANGES_REQUESTED** [review]

File: source/lib-intro.tex [hunk 2]
```diff
@@ -934,35 +961,47 @@
...
-\item the definition is followed by a comment that includes \textit{freestanding}, or
-\item the header synopsis begins with a comment that includes \textit{all freestanding}.
+\item the definition is followed by a comment
+that includes \textit{freestanding}, or
+\item the header synopsis begins with a comment
+that includes \textit{freestanding} and
+the definition is not followed by a comment that includes \textit{hosted}.
 \end{itemize}
-
-\pnum
```
[the remark this answers is not in the document]
Comment (jensmaurer) [reply source/lib-intro.tex:941]: Fixed.

File: source/lib-intro.tex [hunk 3]
```diff
@@ -934,35 +961,47 @@
...
 \end{example}
 
+\pnum
 \begin{example}
 \begin{codeblock}
-// all freestanding
-namespace std {
+#define NULL @\seebelow@      // freestanding
 \end{codeblock}
 \end{example}
```
[the remark this answers is not in the document]
Comment (jensmaurer) [reply source/lib-intro.tex]: Fixed.

File: source/lib-intro.tex [hunk 4]
```diff
@@ -934,35 +961,47 @@
...
-it is \impldef{whether a freestanding deleted function is a freestanding item or a deleted function}
-whether each function definition introduced by a freestanding deleted function
-is a freestanding item or a deleted function\iref{dcl.fct.def.delete}.
 \begin{example}
 \begin{codeblock}
-double abs(double j);           // freestanding-deleted
+// all freestanding
+namespace std {
 \end{codeblock}
 \end{example}
```
[the remark this answers is not in the document]
Comment (jensmaurer) [reply source/lib-intro.tex]: Fixed.

File: source/strings.tex [hunk 5]
```diff
@@ -1466,7 +1478,7 @@
 \item
 \tcode{xpos + str.size() <= size()}
 \item
-\tcode{traits::eq(at(xpos + I), str.at(I))} for all elements \tcode{I} of the string referenced by \tcode{str}.
+\tcode{traits::eq(data_[xpos + I], str.[I])} for all elements \tcode{I} of the string referenced by \tcode{str}.
```
[the remark this answers is not in the document]
Comment (jensmaurer) [reply source/strings.tex]: Fixed.

File: source/strings.tex (hunk 1 above)
Comment (jensmaurer) [reply source/strings.tex]: Fixed.

**Review by JohelEGP — APPROVED** [review]

**Comment by tkoeppe** [comment]
@jensmaurer, @jwakely  is changing `value()` to `**this` still useful/beneficial after 199ed1891d13ea9418a3d352c8b7c678cd879533 / "LWG3973 Monadic operations should be ADL-proof" / https://cplusplus.github.io/LWG/lwg-defects.html#3973?

File: source/support.tex [hunk 6]
```diff
@@ -635,6 +635,8 @@
 #define @\defnlibxname{cpp_lib_format_ranges}@                     202207L // also in \libheader{format}
 #define @\defnlibxname{cpp_lib_formatters}@                        202302L // also in \libheader{stacktrace}, \libheader{thread}
 #define @\defnlibxname{cpp_lib_forward_like}@                      202207L // freestanding, also in \libheader{utility}
+#define @\defnlibxname{cpp_lib_freestanding_algorithm}@            202311L // freestanding, also in \libheader{algorithm}
```
Comment (tkoeppe) [inline source/support.tex]: @jensmaurer, @jwakely Please note that I'm reconciling this with Motion 2 by adding `<numeric>` here, too, for the saturation arithmetic.

**Comment by tkoeppe** [comment]
> @jensmaurer, @jwakely is changing `value()` to `**this` still useful/beneficial after [199ed18](https://github.com/cplusplus/draft/commit/199ed1891d13ea9418a3d352c8b7c678cd879533) / "[LWG3973](https://wg21.link/lwg3973) Monadic operations should be ADL-proof" / https://cplusplus.github.io/LWG/lwg-defects.html#3973?


Hm, why is there even a separate commit for the `**this` changes in [optional.monadic]? Those changes are part of the paper. I'll squash them.

File: source/support.tex (hunk 6 above)
Comment (tkoeppe) [reply source/support.tex]: @jwakely OK, we should make a decision here. Adding a new macro for `numeric` seems reasonable, too, shall I go ahead with that?

Comment (jwakely) [reply source/support.tex]: Yes, I think so. I've pinged Barry and Ben on the PR for that change, for extra eyes.

**Comment by jwakely** [comment]
> @jensmaurer, @jwakely is changing `value()` to `**this` still useful/beneficial after [199ed18](https://github.com/cplusplus/draft/commit/199ed1891d13ea9418a3d352c8b7c678cd879533) / "[LWG3973](https://wg21.link/lwg3973) Monadic operations should be ADL-proof" / https://cplusplus.github.io/LWG/lwg-defects.html#3973?

Unfortunately, 3973 was badly broken, see [LWG 4015](https://cplusplus.github.io/LWG/issue4015). `**this` would be less broken.

But in terms of what's actually been approved for the draft, I think 3973 takes precedence, despite being broken :-( 3973 was explicitly intended to "fix" the changes made here to avoid ADL for `operator*`., even though that's a complete non-issue in most real code, whereas what 3973 requires won't even compile for many real cases.

**Comment by tkoeppe** [comment]
@jwakely So shoud we change `**this` back to `val_` for now?

**Comment by tkoeppe** [comment]
Sending a follow-up commit now.

**Pull request merged**